x/ecocredit/server/marketplace/buy.go

```diff
@@ -2,10 +2,276 @@ package marketplace
+		bal := k.bankKeeper.GetBalance(sdkCtx, buyerAcc, order.BidPrice.Denom)
+		if bal.IsLT(*order.BidPrice) {
+			return nil, sdkerrors.ErrInsufficientFunds.Wrapf("cannot bid %v with a balance of %v", bal, *order.BidPrice)
+		}
+
+		switch selection := order.Selection.Sum.(type) {
+		case *v1.MsgBuy_Order_Selection_SellOrderId:
+			sellOrder, err := k.stateStore.SellOrderTable().Get(ctx, selection.SellOrderId)
+			if err != nil {
+				return nil, fmt.Errorf("sell order %d: %w", selection.SellOrderId, err)
+			}
+			if sellOrder.DisableAutoRetire != order.DisableAutoRetire {
+				return nil, sdkerrors.ErrInvalidRequest.Wrapf("auto-retire mismatch: sell order set to %t, buy "+
+					"order set to %t", sellOrder.DisableAutoRetire, order.DisableAutoRetire)
+			}
+			batch, err := k.coreStore.BatchInfoTable().Get(ctx, sellOrder.BatchId)
+			if err != nil {
+
+				return nil, sdkerrors.ErrIO.Wrapf("error getting batch id %d: %s", sellOrder.BatchId, err.Error())
+			}
+			ct, err := utils.GetCreditTypeFromBatchDenom(ctx, k.coreStore, k.paramsKeeper, batch.BatchDenom)
+			if err != nil {
+				return nil, err
+			}
+			creditOrderQty, err := math.NewPositiveFixedDecFromString(order.Quantity, ct.Precision)
+			if err != nil {
+				return nil, err
+			}
+
+			// check that bid price and ask price denoms match
+			market, err := k.stateStore.MarketTable().Get(ctx, sellOrder.MarketId)
+			if err != nil {
+				return nil, fmt.Errorf("market id %d: %w", sellOrder.MarketId, err)
+			}
+			if order.BidPrice.Denom != market.BankDenom {
+				return nil, sdkerrors.ErrInvalidRequest.Wrapf("bid price denom does not match ask price denom: "+
+					" %s, expected %s", order.BidPrice.Denom, market.BankDenom)
+			}
+			// check that bid price >= sell price
+			sellOrderPricePerCredit, ok := sdk.NewIntFromString(sellOrder.AskPrice)
+			if !ok {
+				return nil, fmt.Errorf("could not convert %s to %T", sellOrder.AskPrice, sdk.Int{})
+			}
+			sellOrderPriceCoin := sdk.Coin{Denom: market.BankDenom, Amount: sellOrderPricePerCredit}
+			if sellOrderPricePerCredit.GT(order.BidPrice.Amount) {
+				return nil, ErrBidTooLow.Wrapf("sell order ask: %v, bid: %v", sellOrderPriceCoin, order.BidPrice)
+			}
+
+			// check address has the total cost (price per * order quantity)
+			cost, err := getTotalCost(sellOrderPricePerCredit, order.Quantity)
+			if err != nil {
+				return nil, err
+			}
+			coinCost := sdk.Coin{Amount: cost, Denom: market.BankDenom}
+			if bal.IsLT(coinCost) {
+				return nil, sdkerrors.ErrInsufficientFunds.Wrapf("requested to purchase %s credits @ %s%s per "+
+					"credit (total %v) with a balance of %v", order.Quantity, sellOrder.AskPrice, market.BankDenom, coinCost, bal)
+			}
+
+			// fill the order, updating balances and the sell order in state
+			if err = k.fillOrder(ctx, sellOrder, buyerAcc, creditOrderQty, coinCost, false); err != nil {
+				return nil, fmt.Errorf("error updating balances: %w", err)
+			}
+			if !sellOrder.DisableAutoRetire {
```

Comment:
i think these events should be emitted in `fillOrder` not here

Reply:
moved https://github.com/regen-network/regen-ledger/pull/899/commits/8f227d75120e9958e206fdaaec7f4455c631c2be#diff-afd3bb1edae672e36e07246bf68948fa67930945a0cfb3b5293d1d1a3707fcdbR202-R239